proposals/0000-fine-grained-unused-warnings.rst

```diff
@@ -0,0 +1,223 @@
+Fine-Grained Unused Warnings
+============================
+
+.. author:: Jakob Brünker
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/434>`_.
+.. contents::
+
+This proposal aims to make warnings about unused names more precise.
+
+Motivation
+----------
+
+::
+
+  a :: Int
+  a = 4
+    where bar = foo
+          foo = 3
+
+Compiling the above code with ``-Wall`` results in the
+these warnings:
+
+::
+
+      Defined but not used: `bar'
+      Defined but not used: `foo'
+
+Sometimes, I look at a warning like this and think "I could have sworn I used
+``foo`` somewhere" - but of course, what the warning is trying to say isn't
+actually that ``foo`` is unused, but rather that all of ``foo``'s usage sites
+are in unused bindings.
+
+Furthermore, while it's useful to see such a warning for CI, to immediately
+know all of the bindings that have to be removed to eliminate the warnings,
+during development, these additional warnings take up a lot of valuable
+screen real estate in ghcid.
+
+Proposed Change Specification
+-----------------------------
+
+The proposed change aims to distinguish between genuinely (or directly) unused bindings and transitively (or indirectly) unused bindings. While the warnings for directly unused bindings remain unchanged, the warnings for indirectly unused bindings will now be controlled by a new flag, ``-freport-indirectly-unused-bindings``.
+
+**Criteria for Transitively Unused Bindings:**
+
+1. **Directly Unused Bindings:** A binding is considered directly unused if it is not referenced in any part of the code. These bindings will generate the usual warnings as before.
+
+2. **Transitively Unused Bindings:** A binding is classified as transitively unused if it is used only within other (directly or indirectly) unused bindings. The warning for these bindings will be reported only if ``-freport-indirectly-unused-bindings` and the relevant existing warning flags (e.g., ``-Wunused-top-binds``, ``-Wunused-local-binds``) are enabled.
```

Comment:
What default value do you specify for `-freport-indirectly-unused-bindings`? My (rather strong) bias is to have it on by default.

(it does feel a little weird to me that this be a `-f` option, but maybe it's the best choice anyway)

Reply:
I'm definitely open to changing the flag prefix, although I'm not sure what would be better